src/distilabel/pipeline/templates/dataset_instruction.py

```diff
@@ -0,0 +1,148 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional
+
+from distilabel.distiset import Distiset
+from distilabel.llms import LLM, InferenceEndpointsLLM
+from distilabel.pipeline import Pipeline
+from distilabel.steps import ExpandColumns, KeepColumns
+from distilabel.steps.tasks import TextGeneration
+
+MODEL = "meta-llama/Meta-Llama-3.1-8B-Instruct"
+from distilabel.models import InferenceEndpointsLLM
+from distilabel.steps.tasks import SelfInstruct
+
+
+class DatasetInstructionResponsePipeline:
+    """Generates instructions and responses for a dataset with input documents.
+
+    This example pipeline can be used for a Supervised Fine-Tuning dataset which you
+    could use to train or evaluate a model. The pipeline generates instructions using the
+    SelfInstruct step and TextGeneration step.
+
+    References:
+        - [Self-Instruct: Aligning Language Models with Self-Generated Instructions](https://arxiv.org/abs/2212.10560)
+
+    Example:
+
+        Generate instructions and responses for a given system prompt:
+
+        ```python
+        from datasets import Dataset
+        from distilabel.pipeline import DatasetInstructionResponsePipeline
+
+        pipeline = DatasetInstructionResponsePipeline(num_instructions=5)
+
+        distiset = pipeline.pipeline.run(
+            use_cache=False,
+            dataset=Dataset.from_list(
+                mapping=[
+                    {
+                        "input": "<document>",
+                    }
+                ]
+            ),
+        )
+        ```
+    """
+
+    def __init__(
+        self,
+        llm: Optional[LLM] = None,
+        system_prompt: str = "You are a creative AI Assistant writer.",
+        hf_token: Optional[str] = None,
+        num_instructions: int = 2,
+        batch_size: int = 1,
+    ) -> None:
+        if llm is None:
+            self.llm: LLM = InferenceEndpointsLLM(
+                model_id=MODEL,
+                tokenizer_id=MODEL,
+                magpie_pre_query_template="llama3",
+                generation_kwargs={
+                    "temperature": 0.9,
+                    "do_sample": True,
+                    "max_new_tokens": 2048,
+                    "stop_sequences": [
+                        "<|eot_id|>",
+                        "<|start_header_id|>",
+                        "assistant",
+                        " \n\n",
+                    ],
+                },
+                api_key=hf_token,
+            )
+        else:
+            self.llm = llm
+
+        self.pipeline: Pipeline = self._get_pipeline(
+            system_prompt=system_prompt,
+            num_instructions=num_instructions,
+            batch_size=batch_size,
+        )
+
+    def run(self, dataset, **kwargs) -> Distiset:
```

Comment:
perhaps we should define the additional arguments a bit more explicit? `num_instructions` etc. We might even consider aligning it more with the other pipeline and use a mapping from `num_instructions` to `num_rows`.

Reply:
I don't follow this properly sorry. The other pipeline use num_rows and n_turns because it's based on magpie. Are you suggesting that we calculate a num_rows based on num_instructions * input dataset rows?

Reply:
@burtenshaw, it is a bit related to this comment but we can skip it at this moment. I was suggesting the argument for the number of samples to generate could have benefited from some unification because `num_samples`, `num_instruction` and `num_rows` have the same meaning to the users. We do the same in the synthetic data generator to simplify all of the various arguments.

https://github.com/argilla-io/distilabel/pull/1076#issuecomment-2531082774

Reply:
Ok. I agree it would make sense to have something unified.